spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java

```diff
@@ -0,0 +1,156 @@
+import com.google.cloud.bigquery.storage.v1.FinalizeWriteStreamResponse;
+import com.google.cloud.bigquery.storage.v1.JsonStreamWriter;
+import com.google.cloud.bigquery.storage.v1.TableName;
+import com.google.cloud.bigquery.storage.v1.WriteStream;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.Descriptors.DescriptorValidationException;
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Phaser;
+import javax.annotation.concurrent.GuardedBy;
+import org.json.JSONArray;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Helper class for using BigQuery storage write API in exactly once delivery mode */
+public class BigQueryJsonDataWriter implements AutoCloseable {
+
+  private final JsonStreamWriter streamWriter;
+  // Track the number of in-flight requests to wait for all responses before shutting down.
+  private final Phaser inflightRequestCount = new Phaser(1);
+
+  private final Object lock = new Object();
+
+  private final Logger logger = LoggerFactory.getLogger(BigQueryJsonDataWriter.class);
+
+  @GuardedBy("lock")
+  private RuntimeException error = null;
+
+  /**
+   * @param parentTable against which the writer has to be initialized
+   * @param client BigQueryWriteClient reference which has to be used for writing to the database
+   */
+  public BigQueryJsonDataWriter(TableName parentTable, BigQueryWriteClient client)
+      throws DescriptorValidationException, IOException, InterruptedException {
+    // Initialize a write stream for the specified table.
+    // For more information on WriteStream.Type, see:
+    // https://googleapis.dev/java/google-cloud-bigquerystorage/latest/com/google/cloud/bigquery/storage/v1/WriteStream.Type.html
+    WriteStream stream = WriteStream.newBuilder().setType(WriteStream.Type.PENDING).build();
+
+    CreateWriteStreamRequest createWriteStreamRequest =
+        CreateWriteStreamRequest.newBuilder()
+            .setParent(parentTable.toString())
+            .setWriteStream(stream)
+            .build();
+    WriteStream writeStream = client.createWriteStream(createWriteStreamRequest);
+
+    // Use the JSON stream writer to send records in JSON format.
+    // For more information about JsonStreamWriter, see:
+    // https://googleapis.dev/java/google-cloud-bigquerystorage/latest/com/google/cloud/bigquery/storage/v1beta2/JsonStreamWriter.html
+    streamWriter =
+        JsonStreamWriter.newBuilder(writeStream.getName(), writeStream.getTableSchema()).build();
+  }
+
+  /**
+   * @param data JSONArray to be appended
+   * @param offset offset at which data has to be added
+   */
+  public void append(JSONArray data, long offset)
+      throws DescriptorValidationException, IOException, ExecutionException {
+    synchronized (this.lock) {
+      // If earlier appends have failed, we need to reset before continuing.
+      if (this.error != null) {
+        throw this.error;
```

Comment:
Should the error get reset based on comment above? It seems to be unrecoverable based on usage in the template.

Reply:
We get to know about the failure asynchronously @ `onFailure` and we throw this exception when the developer tries to append @ `BigQueryTemplate`. I have captured a task in the design document to better handle the errors (on the lines of what we had discussed, like returning the unique set of exceptions and the count of exceptions). Is it okay if we plan to address this in a follow-up PR ?